dist/api_config.py

```diff
@@ -498,9 +616,157 @@ def getsubconfigstr(c):
+            structs += '{}uint64_t {};\n'.format(indent, get_allowable_name(name))
+            inits += '{}{},\n'.format(indent, build_key_initializer(subnames))
+        else:
+            lbrace = '{'
+            rbrace = '}'
+            structs += '{}struct {}\n'.format(indent, lbrace)
+            inits += '{}{}\n'.format(indent, lbrace)
+            (s2, i2) = gen_conf_key_struct_init(indent + '  ', subnames, h)
+            structs += s2 + '{}{} {};\n'.format(indent, rbrace, name)
+            inits += i2 + '{}{},\n'.format(indent, rbrace)
+    return [structs, inits]
+
+def get_conf_counts(configs):
+    global config_num, config_key
+    nconf = 1
+    nitem = 0
+    for c in configs:
+        ty = gettype(c)
+        nitem += 1
+        if ty == 'category':
+            (subconf, subitem) = get_conf_counts(c.subconfig)
+            nconf += subconf
+            nitem += subitem
+    return (nconf, nitem)
+
+# Update the config.h and conf.h files with the #defines for the configuration entries.
 if not test_config:
     config_h = ModifyFile('../src/include/config.h')
+    conf_h = ModifyFile('../src/include/conf.h')
+
     with config_h.replace_fragment('configuration section') as tfile:
-            tfile.write(config_defines)
+        tfile.write(config_defines)
+        tfile.write('\n')
+        tfile.write('extern const char __WT_CONFIG_CHOICE_NULL[]; /* not set in configuration */\n')
+        for choice in sorted(choices_values):
+            tfile.write('extern const char __WT_CONFIG_CHOICE_{}[];\n'.format(choice))
+
+    conf_keys = dict()
+
+    add_conf_keys(api_data_def.methods, conf_keys, True)
+
+    # Assign unique numbers to all keys used in configuration, regardless of "level".
+    with conf_h.replace_fragment('API configuration keys') as tfile:
+        count = 0
+        for name in sorted(keynumber.numbering.keys()):
+            off = keynumber.get(name)
+            tfile.write('#define WT_CONF_ID_{} {}ULL\n'.format(name, off))
+            count += 1
+        assert count == keynumber.count()
+        tfile.write('\n#define WT_CONF_ID_COUNT {}\n'.format(count))
+
+    with conf_h.replace_fragment('Configuration key structure') as tfile:
+        (structs, inits) = gen_conf_key_struct_init('    ', [], conf_keys)
+        tfile.write('static const struct {\n')
+        tfile.write(structs)
+        tfile.write('} WT_CONF_ID_STRUCTURE = {\n')
+        tfile.write(inits)
+        tfile.write('};\n')
+
+    with conf_h.replace_fragment('Per-API configuration structure declarations') as tfile:
+        method_to_counts = dict()
+        for name, method in api_data_def.methods.items():
+            config = method.config
+            if config:
+                method_to_counts[name] = get_conf_counts(config)
+
+        count = 0
+        for name in sorted(api_data_def.methods.keys()):
+            # Extract class and method name
+            (clname, mname) = get_method_name_parts(name)
+            count += 1
```

Comment:
The code is misleading (even if correct), it should print the count of number of elements in the array.  So I changed it to do that directly, rather than have a counter.  I added a comment too.  Thanks for spotting that.